contributors/design-proposals/volume-provisioning-configuration.md

```diff
@@ -0,0 +1,207 @@
+  * We will update examples for all supported internal provisioners with selectorOptions for easier copy/paste.
+  * `kubectl describe storageclass` will show available options ad values in a user-friendly form.
+
+* There is no code in Kubernetes that validates user supplied `ConfigMap` with `allowedProvisioningOptions` in StorageClasses. Users may add invalid option keys or invalid values to their `ConfigMaps` and it's up to the individual provisioner to validate the `ConfigMap` and throw appropriate errors.
+
+## Examples
+
+### Ultimate StorageClass for AWS
+This example shows a rich StorageClass with three user-tunable parameters.
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: StorageClass
+metadata:
+  name: slow
+provisioner: kubernetes.io/aws-ebs
+parameters:
+  type: gp2
+  zones: us-east-1a, us-east-1b, us-east-1c, us-east-2a, us-east-2b, us-east-2c
+allowedProvisioningOptions:
+  - key: zone
+    allowedValues:
+        - us-east-1a
+        - us-east-1b
+        - us-east-1c
+        - us-east-2a
+        - us-east-2b
+        - us-east-2c
+    description: "Requested AWS availability zone"
+
+  - key: encryption
+    defaultValue: "false"
+    allowedValues:
+      - "true"
+      - "false"
+    description: "Request volume encryption"
+
+  - key: kmsKeyId
+    description: "ID of AWS encryption key"
+```
+
+kubectl output:
+```shell
+$ kubectl describe storageclass default
+Name:           default
+IsDefaultClass: Yes
+Provisioner:    kubernetes.io/aws-ebs
+AllowedProvisioningOptions:
+    zone
+      - requested AWS availability zone
+      - values:
+        - us-east-1a
+        - us-east-1b
+        - us-east-1c
+        - us-east-2a
+        - us-east-2b
+        - us-east-2c
+
+    encryption
+      - request volume encryption
+      - values:
+        - true
+        - false (default)
+
+    kmsKeyId
+      - ID of AWS encryption key
+
+Parameters: type=pd-standard,zones=us-east-1a, us-east-1b, us-east-1c, us-east-2a, us-east-2b, us-east-2c
+
+```
+
+* All options are optional.
+* There is no default zone selected, provisioner may choose any zone if PVC's ConfigMap does not require any.
+* Encryption is off by default.
+
+Regular user first cretes a `ConfigMap` with chosen provisioning options:
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-provisioning-config
+data:
+  zone: us-east-1d
+  type: gp2
+  encrypted: "true"
+  kmsKeyId: arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012
+```
+
+Then the user can create any number of PVCs referring to this `ConfigMap`:
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: myclaim
+spec:
+  class: slow
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1500Mi
+  provisioningOptions:
+    name: my-provisioning-config
+```
+
+* Note that this ConfigMap is used only during provisioning. Changes in the ConfigMap are not reflected in already provisioned PVs.
```

Comment:
The idea is that most people won't care about configuration of storage at all. Admin prepares a handful of storage classes for generic usage and user pick one of them and that's it. ConfigMap is for those who need specific options for their data. This proposal assumes there will be a limited amount of those, otherwise it gets indeed complicated.